File: BackEnd/src/domain/dtos/user/update-preferences-request.dto.ts
/**
 * @file update-preferences-request.dto.ts
 * @description DTO para actualizar las preferencias de notificación y comportamiento del usuario.
 * Todos los campos son opcionales: solo se actualizan los que se envían.
 * @module User
 * @layer Domain
 */

import { IsBoolean, IsOptional } from 'class-validator';

/**
 * Campos de preferencias actualizables.
 * Coincide con la interfaz UserPreferences de la entidad User.
 */
export class UpdatePreferencesRequestDto {
  /** Aparecer en búsquedas de chat de otros usuarios */
  @IsOptional() @IsBoolean() appearInChatSearch?: boolean;

  /** Considerar el clima al calcular recordatorios de riego por defecto */
  @IsOptional() @IsBoolean() considerWeatherByDefault?: boolean;

  /** Si true, el perfil no aparece en el feed público ni acepta nuevas conversaciones */
  @IsOptional() @IsBoolean() isPrivate?: boolean;

  /**
   * Si false, el backend no envía push FCM aunque el usuario tenga
   * fcmToken registrado. Las capas in-app y socket no se ven afectadas.
   */
  @IsOptional() @IsBoolean() pushNotifications?: boolean;
}

File: FrontEnd/lib/domain/dtos/user/update_preferences_request_dto.dart
/// @file update_preferences_request_dto.dart
/// @description DTO para actualizar las preferencias del usuario.
/// Todos los campos son opcionales — solo se envían los que cambian.
/// @module User
/// @layer Domain
library;

// ═══════════════════════════════════════════════════════════════════════════════
// UPDATE PREFERENCES REQUEST DTO
// ═══════════════════════════════════════════════════════════════════════════════

/// DTO de actualización de preferencias del usuario.
///
/// Todos los campos son opcionales. El backend aplica patch parcial.
///
/// El DTO no incluye el campo `language`: el idioma de la app vive solo
/// en el cliente (i18n esqueleto), no se persiste en el perfil del
/// usuario.
class UpdatePreferencesRequestDto {
  const UpdatePreferencesRequestDto({
    this.pushNotifications,
    this.profilePublic,
  });

  /// Habilita o deshabilita notificaciones push.
  final bool? pushNotifications;

  /// Hace el perfil público o privado en la comunidad.
  final bool? profilePublic;

  Map<String, dynamic> toJson() => {
    if (pushNotifications != null) 'pushNotifications': pushNotifications,
    // El backend usa isPrivate (inverso de profilePublic).
    if (profilePublic     != null) 'isPrivate':         !profilePublic!,
  };
}
